src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonValueTrimmable.cs

```diff
@@ -51,5 +51,97 @@ public override void WriteTo(Utf8JsonWriter writer, JsonSerializerOptions? optio
                 JsonSerializer.Serialize(writer, _value, _jsonTypeInfo);
             }
         }
+
+        internal override JsonNode InternalDeepClone()
+        {
+            if (_converter is not null)
+            {
+                return _value is JsonElement element
+                    ? new JsonValueTrimmable<JsonElement>(element.Clone(), JsonMetadataServices.JsonElementConverter, Options)
+                    : new JsonValueTrimmable<TValue>(_value, _converter, Options);
+            }
+            else
+            {
+                Debug.Assert(_jsonTypeInfo != null);
+                return JsonSerializer.SerializeToNode(_value, _jsonTypeInfo)!;
+            }
+        }
+
+        internal override bool DeepEquals(JsonNode? node)
+        {
+            if (node is null)
+            {
+                return false;
+            }
+
+            if (_jsonTypeInfo is not null)
+            {
+                JsonNode? jsonNode = JsonSerializer.SerializeToNode(_value, _jsonTypeInfo);
+                return DeepEquals(jsonNode, node);
+            }
+            else
+            {
+                if (node is JsonArray || node is JsonObject)
+                {
+                    return false;
+                }
+
+                if (node is JsonValueTrimmable<JsonElement> jsonElementNodeOther && _value is JsonElement jsonElementCurrent)
+                {
+                    if (jsonElementNodeOther._value.ValueKind != jsonElementCurrent.ValueKind)
+                    {
+                        return false;
+                    }
+
+                    switch (jsonElementCurrent.ValueKind)
+                    {
+                        case JsonValueKind.String:
+                        case JsonValueKind.Number:
+                        case JsonValueKind.True:
+                        case JsonValueKind.False:
+                            return jsonElementCurrent.GetRawValue().Span.SequenceEqual(jsonElementNodeOther._value.GetRawValue().Span);
+                        default:
```

Comment:
This will always return false for `JsonValueKind.Object` and `Array`. Is this correct?

Reply:
Shouldn't there be an equivalent `DeepEquals` implementation in `JsonElement` that we should be calling here?

Reply:
> This will always return false for `JsonValueKind.Object` and `Array`. Is this correct?

yes @eiriktsarpalis

Reply:
> Shouldn't there be an equivalent `DeepEquals` implementation in `JsonElement` that we should be calling here?

I don't think we should have because this is a special context (`JsonElement` in this `JsonValueTrimmable` class will never have `JsonValueKind` as `Object` or `Array` if we introduce a new method `DeepEquals` inside `JsonElement` we need to handle `Object` and `Array` case and that will introduce another thing), how about your idea ? @eiriktsarpalis

Reply:
> That seems to violate the notion of deep equality though.

@eiriktsarpalis  as you can see in this line of code (https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonValue.CreateOverloads.cs#L321) `JsonElement` constructed in `JsonValueTrimmable` will never have `JsonValueKind` as `Object` or `Array`.

Reply:
> JsonElement in this JsonValueTrimmable class will never have JsonValueKind as Object or Array

Maybe I'm missing something, but it seems I can do that with `JsonValueTrimmable`:

```C#
JsonElement element = JsonDocument.Parse("[1,2,3]").RootElement;
JsonValue? node = JsonValue.Create(element, MyContext.Default.JsonElement);

[JsonSerializable(typeof(JsonElement))]
public partial class MyContext : JsonSerializerContext
{
}
```

Reply:
Ah, I see. It seems we're throwing an exception on construction. In that case we're ok, but could you add an assert checking that `JsonValueKind` is never one of the two in the equality comparison method?

Reply:
I added the assertion.

Reply:
Maybe just add a `Debug.Fail` in the `default` case?